File: src/security/gpg.zig
const std = @import("std");

pub const TrustLevel = enum {
    undefined,
    never,
    marginal,
    full,
    ultimate,
    
    pub fn toFloat(self: TrustLevel) f32 {
        return switch (self) {
            .undefined => 0.0,
            .never => -2.0,
            .marginal => 0.5,
            .full => 1.5,
            .ultimate => 2.0,
        };
    }
};

pub const GpgVerification = struct {
    is_valid: bool,
    key_id: ?[]const u8,
    fingerprint: ?[]const u8,
    trust_level: TrustLevel,
    signer_name: ?[]const u8,
    signer_email: ?[]const u8,
    creation_time: ?u64,
    expiration_time: ?u64,
    error_message: ?[]const u8,
    
    pub fn getTrustScore(self: *const GpgVerification) f32 {
        if (!self.is_valid) return -1.0;
        return self.trust_level.toFloat();
    }
};

pub const KeyInfo = struct {
    key_id: []const u8,
    fingerprint: []const u8,
    user_ids: [][]const u8,
    trust_level: TrustLevel,
    creation_time: u64,
    expiration_time: ?u64,
    is_revoked: bool,
    is_expired: bool,
    
    pub fn deinit(self: *KeyInfo, allocator: std.mem.Allocator) void {
        allocator.free(self.key_id);
        allocator.free(self.fingerprint);
        for (self.user_ids) |uid| {
            allocator.free(uid);
        }
        allocator.free(self.user_ids);
    }
};

pub const GpgManager = struct {
    allocator: std.mem.Allocator,
    gpg_binary: []const u8,
    gnupg_home: ?[]const u8,
    keyservers: []const []const u8,
    auto_import_keys: bool,
    
    pub fn init(allocator: std.mem.Allocator) !*GpgManager {
        const self = try allocator.create(GpgManager);
        self.* = .{
            .allocator = allocator,
            .gpg_binary = "gpg",
            .gnupg_home = null,
            .keyservers = &.{
                "hkps://keys.openpgp.org",
                "hkps://keyserver.ubuntu.com",
                "hkps://pgp.mit.edu",
            },
            .auto_import_keys = true,
        };
        return self;
    }
    
    pub fn deinit(self: *GpgManager) void {
        if (self.gnupg_home) |home| {
            self.allocator.free(home);
        }
        self.allocator.destroy(self);
    }
    
    pub fn verifySignature(self: *GpgManager, file_path: []const u8, signature_path: []const u8) !GpgVerification {
        var args = std.ArrayList([]const u8){};
        defer args.deinit(self.allocator);
        
        try args.append(self.allocator, self.gpg_binary);
        try args.append(self.allocator, "--verify");
        try args.append(self.allocator, "--status-fd");
        try args.append(self.allocator, "1");
        
        if (self.gnupg_home) |home| {
            try args.append(self.allocator, "--homedir");
            try args.append(self.allocator, home);
        }
        
        try args.append(self.allocator, signature_path);
        try args.append(self.allocator, file_path);
        
        const result = try std.process.Child.run(.{
            .allocator = self.allocator,
            .argv = args.items,
        });
        defer self.allocator.free(result.stdout);
        defer self.allocator.free(result.stderr);
        
        return self.parseGpgOutput(result.stdout, result.stderr, result.term);
    }
    
    pub fn verifyPkgbuildSignature(self: *GpgManager, pkg_dir: []const u8) !GpgVerification {
        const pkgbuild_path = try std.fs.path.join(self.allocator, &.{ pkg_dir, "PKGBUILD" });
        defer self.allocator.free(pkgbuild_path);
        
        const sig_path = try std.fmt.allocPrint(self.allocator, "{s}.sig", .{pkgbuild_path});
        defer self.allocator.free(sig_path);
        
        // Check if signature file exists
        std.fs.accessAbsolute(sig_path, .{}) catch {
            return GpgVerification{
                .is_valid = false,
                .key_id = null,
                .fingerprint = null,
                .trust_level = .undefined,
                .signer_name = null,
                .signer_email = null,
                .creation_time = null,
                .expiration_time = null,
                .error_message = try self.allocator.dupe(u8, "No signature file found"),
            };
        };
        
        var verification = try self.verifySignature(pkgbuild_path, sig_path);
        
        // Auto-import missing keys if verification failed due to missing key
        if (!verification.is_valid and self.auto_import_keys) {
            if (verification.key_id) |key_id| {
                if (self.importKeyFromKeyserver(key_id)) {
                    // Retry verification after importing key
                    if (verification.error_message) |msg| {
                        self.allocator.free(msg);
                    }
                    verification = try self.verifySignature(pkgbuild_path, sig_path);
                } else |_| {
                    // Key import failed, keep original verification result
                }
            }
        }
        
        return verification;
    }
    
    pub fn importKeyFromKeyserver(self: *GpgManager, key_id: []const u8) !void {
        for (self.keyservers) |keyserver| {
            const import_result = self.tryImportFromKeyserver(key_id, keyserver) catch continue;
            if (import_result) return;
        }
        return error.KeyImportFailed;
    }
    
    fn tryImportFromKeyserver(self: *GpgManager, key_id: []const u8, keyserver: []const u8) !bool {
        var args = std.ArrayList([]const u8){};
        defer args.deinit(self.allocator);
        
        try args.append(self.allocator, self.gpg_binary);
        try args.append(self.allocator, "--keyserver");
        try args.append(self.allocator, keyserver);
        try args.append(self.allocator, "--recv-keys");
        
        if (self.gnupg_home) |home| {
            try args.append(self.allocator, "--homedir");
            try args.append(self.allocator, home);
        }
        
        try args.append(self.allocator, key_id);
        
        const result = try std.process.Child.run(.{
            .allocator = self.allocator,
            .argv = args.items,
        });
        defer self.allocator.free(result.stdout);
        defer self.allocator.free(result.stderr);
        
        return result.term == .Exited and result.term.Exited == 0;
    }
    
    pub fn getKeyInfo(self: *GpgManager, key_id: []const u8) !?KeyInfo {
        var args = std.ArrayList([]const u8){};
        defer args.deinit(self.allocator);
        
        try args.append(self.allocator, self.gpg_binary);
        try args.append(self.allocator, "--list-keys");
        try args.append(self.allocator, "--with-colons");
        try args.append(self.allocator, "--fixed-list-mode");
        
        if (self.gnupg_home) |home| {
            try args.append(self.allocator, "--homedir");
            try args.append(self.allocator, home);
        }
        
        try args.append(self.allocator, key_id);
        
        const result = try std.process.Child.run(.{
            .allocator = self.allocator,
            .argv = args.items,
        });
        defer self.allocator.free(result.stdout);
        defer self.allocator.free(result.stderr);
        
        if (result.term != .Exited or result.term.Exited != 0) {
            return null;
        }
        
        return self.parseKeyInfo(result.stdout);
    }
    
    pub fn getTrustLevel(self: *GpgManager, key_id: []const u8) !TrustLevel {
        if (try self.getKeyInfo(key_id)) |key_info| {
            defer key_info.deinit(self.allocator);
            return key_info.trust_level;
        }
        return .undefined;
    }
    
    pub fn autoFetchMissingKeys(self: *GpgManager, signature_error: []const u8) !void {
        // Parse error message to extract key ID
        if (std.mem.indexOf(u8, signature_error, "NO_PUBKEY")) |start| {
            const key_start = start + 10; // Length of "NO_PUBKEY "
            if (key_start < signature_error.len) {
                var key_end = key_start;
                while (key_end < signature_error.len and std.ascii.isAlphanumeric(signature_error[key_end])) {
                    key_end += 1;
                }
                
                if (key_end > key_start) {
                    const key_id = signature_error[key_start..key_end];
                    try self.importKeyFromKeyserver(key_id);
                }
            }
        }
    }
    
    fn parseGpgOutput(self: *GpgManager, stdout: []const u8, stderr: []const u8, term: std.process.Child.Term) !GpgVerification {
        var verification = GpgVerification{
            .is_valid = false,
            .key_id = null,
            .fingerprint = null,
            .trust_level = .undefined,
            .signer_name = null,
            .signer_email = null,
            .creation_time = null,
            .expiration_time = null,
            .error_message = null,
        };
        
        if (term != .Exited or term.Exited != 0) {
            verification.error_message = try self.allocator.dupe(u8, stderr);
            return verification;
        }
        
        // Parse GPG status output
        var lines = std.mem.tokenizeScalar(u8, stdout, '\n');
        while (lines.next()) |line| {
            if (std.mem.startsWith(u8, line, "[GNUPG:] GOODSIG")) {
                verification.is_valid = true;
                // Extract key ID and signer info
                var parts = std.mem.tokenizeScalar(u8, line, ' ');
                _ = parts.next(); // Skip "[GNUPG:]"
                _ = parts.next(); // Skip "GOODSIG"
                
                if (parts.next()) |key_id| {
                    verification.key_id = try self.allocator.dupe(u8, key_id);
                }
                
                // Rest is signer name/email
                var signer_parts = std.ArrayList(u8){};
                defer signer_parts.deinit(self.allocator);
                
                while (parts.next()) |part| {
                    if (signer_parts.items.len > 0) {
                        try signer_parts.append(self.allocator, ' ');
                    }
                    try signer_parts.appendSlice(self.allocator, part);
                }
                
                if (signer_parts.items.len > 0) {
                    verification.signer_name = try signer_parts.toOwnedSlice(self.allocator);
                }
            } else if (std.mem.startsWith(u8, line, "[GNUPG:] TRUST_")) {
                // Extract trust level
                if (std.mem.indexOf(u8, line, "TRUST_ULTIMATE")) |_| {
                    verification.trust_level = .ultimate;
                } else if (std.mem.indexOf(u8, line, "TRUST_FULL")) |_| {
                    verification.trust_level = .full;
                } else if (std.mem.indexOf(u8, line, "TRUST_MARGINAL")) |_| {
                    verification.trust_level = .marginal;
                } else if (std.mem.indexOf(u8, line, "TRUST_NEVER")) |_| {
                    verification.trust_level = .never;
                }
            } else if (std.mem.startsWith(u8, line, "[GNUPG:] BADSIG")) {
                verification.is_valid = false;
                verification.error_message = try self.allocator.dupe(u8, "Bad signature");
            } else if (std.mem.startsWith(u8, line, "[GNUPG:] NO_PUBKEY")) {
                verification.is_valid = false;
                verification.error_message = try self.allocator.dupe(u8, "Public key not available");
                
                // Extract missing key ID
                var parts = std.mem.tokenizeScalar(u8, line, ' ');
                _ = parts.next(); // Skip "[GNUPG:]"
                _ = parts.next(); // Skip "NO_PUBKEY"
                
                if (parts.next()) |key_id| {
                    verification.key_id = try self.allocator.dupe(u8, key_id);
                }
            }
        }
        
        return verification;
    }
    
    fn parseKeyInfo(self: *GpgManager, output: []const u8) !?KeyInfo {
        var lines = std.mem.tokenizeScalar(u8, output, '\n');
        var key_info: ?KeyInfo = null;
        var user_ids = std.ArrayList([]const u8){};
        
        while (lines.next()) |line| {
            var fields = std.mem.tokenizeScalar(u8, line, ':');
            const record_type = fields.next() orelse continue;
            
            if (std.mem.eql(u8, record_type, "pub")) {
                // Public key record
                _ = fields.next(); // Skip validity
                _ = fields.next(); // Skip key length
                _ = fields.next(); // Skip algorithm
                
                const key_id = fields.next() orelse continue;
                const creation_date = fields.next() orelse "0";
                const expiration_date = fields.next();
                
                key_info = KeyInfo{
                    .key_id = try self.allocator.dupe(u8, key_id),
                    .fingerprint = try self.allocator.dupe(u8, key_id), // Simplified
                    .user_ids = &.{},
                    .trust_level = .undefined,
                    .creation_time = std.fmt.parseInt(u64, creation_date, 10) catch 0,
                    .expiration_time = if (expiration_date) |exp| std.fmt.parseInt(u64, exp, 10) catch null else null,
                    .is_revoked = false,
                    .is_expired = false,
                };
            } else if (std.mem.eql(u8, record_type, "uid") and key_info != null) {
                // User ID record
                _ = fields.next(); // Skip validity
                _ = fields.next(); // Skip creation date
                _ = fields.next(); // Skip expiration date
                _ = fields.next(); // Skip hash
                _ = fields.next(); // Skip class
                
                const user_id = fields.next() orelse continue;
                try user_ids.append(self.allocator, try self.allocator.dupe(u8, user_id));
            }
        }
        
        if (key_info) |*info| {
            info.user_ids = try user_ids.toOwnedSlice();
            return info.*;
        }
        
        // Clean up if no key found
        for (user_ids.items) |uid| {
            self.allocator.free(uid);
        }
        user_ids.deinit(self.allocator);
        
        return null;
    }
};

File: src/gpg.rs
use std::path::Path;
use std::process::Command;
use tokio::process::Command as TokioCommand;

/// Result of GPG signature verification using --status-fd parsing
#[derive(Debug, Clone)]
pub enum VerificationResult {
    /// Signature is valid and key is trusted
    Valid { key_id: String, fingerprint: String },
    /// Signature is cryptographically valid but key is not trusted
    ValidUntrusted { key_id: String, fingerprint: String },
    /// Signature is invalid (bad signature)
    Invalid {
        #[allow(dead_code)]
        reason: String,
    },
    /// Key needed to verify is missing from keyring
    MissingKey { key_id: String },
    /// No signature file exists
    NoSignature,
    /// Error running GPG
    Error(#[allow(dead_code)] String),
}

impl VerificationResult {
    /// Returns true only if signature is cryptographically valid (trusted or not)
    #[allow(dead_code)]
    pub fn is_valid(&self) -> bool {
        matches!(
            self,
            VerificationResult::Valid { .. } | VerificationResult::ValidUntrusted { .. }
        )
    }

    /// Returns true if signature is valid AND key is trusted
    #[allow(dead_code)]
    pub fn is_trusted(&self) -> bool {
        matches!(self, VerificationResult::Valid { .. })
    }
}

/// Verify a detached signature using GPG --status-fd for structured output
pub fn verify_signature(sig_path: &Path, file_path: &Path) -> VerificationResult {
    if !sig_path.exists() {
        return VerificationResult::NoSignature;
    }
    if !file_path.exists() {
        return VerificationResult::Error("File to verify does not exist".to_string());
    }

    // Use --status-fd 1 to get machine-readable status on stdout
    let output = Command::new("gpg")
        .args(["--status-fd", "1", "--verify"])
        .arg(sig_path)
        .arg(file_path)
        .output();

    match output {
        Ok(out) => parse_gpg_status(&out.stdout, &out.stderr),
        Err(e) => VerificationResult::Error(format!("Failed to run gpg: {}", e)),
    }
}

/// Parse GPG --status-fd output for verification result
fn parse_gpg_status(stdout: &[u8], stderr: &[u8]) -> VerificationResult {
    let status = String::from_utf8_lossy(stdout);
    let errors = String::from_utf8_lossy(stderr);

    let mut key_id = String::new();
    let mut fingerprint = String::new();
    let mut good_sig = false;
    let mut bad_sig = false;
    let mut no_pubkey = false;
    let mut trust_level = "UNDEFINED";

    // Parse status lines (format: [GNUPG:] KEYWORD args...)
    for line in status.lines() {
        if !line.starts_with("[GNUPG:]") {
            continue;
        }
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() < 2 {
            continue;
        }

        match parts[1] {
            "GOODSIG" => {
                good_sig = true;
                if parts.len() > 2 {
                    key_id = parts[2].to_string();
                }
            }
            "BADSIG" => {
                bad_sig = true;
                if parts.len() > 2 {
                    key_id = parts[2].to_string();
                }
            }
            "ERRSIG" => {
                // ERRSIG keyid algo hashalgo class time rc
                // rc=9 means missing key
                if parts.len() > 2 {
                    key_id = parts[2].to_string();
                }
                if parts.len() > 7 && parts[7] == "9" {
                    no_pubkey = true;
                }
            }
            "NO_PUBKEY" => {
                no_pubkey = true;
                if parts.len() > 2 {
                    key_id = parts[2].to_string();
                }
            }
            "VALIDSIG" => {
                // VALIDSIG fingerprint creation_date sig_date ... trust_model trust_level
                if parts.len() > 2 {
                    fingerprint = parts[2].to_string();
                }
            }
            "TRUST_UNDEFINED" | "TRUST_NEVER" => trust_level = "UNDEFINED",
            "TRUST_MARGINAL" => trust_level = "MARGINAL",
            "TRUST_FULLY" => trust_level = "FULLY",
            "TRUST_ULTIMATE" => trust_level = "ULTIMATE",
            _ => {}
        }
    }

    // Determine result
    if bad_sig {
        return VerificationResult::Invalid {
            reason: format!("Bad signature from key {}", key_id),
        };
    }

    if no_pubkey {
        return VerificationResult::MissingKey { key_id };
    }

    if good_sig {
        if trust_level == "FULLY" || trust_level == "ULTIMATE" {
            return VerificationResult::Valid {
                key_id,
                fingerprint,
            };
        } else {
            return VerificationResult::ValidUntrusted {
                key_id,
                fingerprint,
            };
        }
    }

    // No clear result - check stderr for clues
    if errors.contains("No public key") {
        // Try to extract key ID from stderr
        for line in errors.lines() {
            if line.contains("key ID")
                && let Some(idx) = line.rfind(' ')
            {
                return VerificationResult::MissingKey {
                    key_id: line[idx + 1..].to_string(),
                };
            }
        }
        return VerificationResult::MissingKey {
            key_id: "unknown".to_string(),
        };
    }

    VerificationResult::Error(format!("Could not parse GPG output: {}", errors))
}

/// Show GPG key info (sync) - debugging/informational output
///
/// This function is for user-facing key inspection, not for trust decisions.
/// For programmatic key trust checking, use `get_trust_level()`.
pub fn show_gpg_key_info(keyid: &str) {
    let output = Command::new("gpg")
        .args(["--list-keys", keyid, "--with-colons"])
        .output();
    if let Ok(out) = output {
        let stdout = String::from_utf8_lossy(&out.stdout);
        for line in stdout.lines() {
            if line.starts_with("pub:") {
                let fields: Vec<&str> = line.split(':').collect();
                if fields.len() > 6 {
                    println!(
                        "[reap] gpg :: Key trust: {} Expiry: {}",
                        fields[1], fields[6]
                    );
                }
            }
        }
    }
}

/// Helper to get GPG trust level for a keyid
pub fn get_trust_level(keyid: &str) -> Option<String> {
    let output = Command::new("gpg")
        .args(["--list-keys", "--with-colons", keyid])
        .output()
        .ok()?;
    let stdout = String::from_utf8_lossy(&output.stdout);
    for line in stdout.lines() {
        if line.starts_with("pub:") || line.starts_with("uid:") {
            let fields: Vec<&str> = line.split(':').collect();
            if fields.len() > 1 {
                let trust = match fields[1] {
                    "f" => "Full",
                    "m" => "Marginal",
                    "n" => "None",
                    "u" => "Ultimate",
                    other => other,
                };
                return Some(trust.to_string());
            }
        }
    }
    None
}

/// GPG PKGBUILD signature check returning structured result
///
/// Use this for programmatic verification where you need to inspect the result.
#[allow(dead_code)] // API for callers who want structured results
pub fn gpg_check_structured(pkgdir: &Path) -> VerificationResult {
    let sig_path = pkgdir.join("PKGBUILD.sig");
    let pkgb_path = pkgdir.join("PKGBUILD");
    verify_signature(&sig_path, &pkgb_path)
}

/// Enhanced GPG PKGBUILD signature check with auto key fetch and console output
///
/// This wraps `verify_signature()` and adds:
/// - Console output for user feedback
/// - Auto key fetch on missing key
/// - Retry after key import
pub fn gpg_check(pkgdir: &Path) -> Result<(), String> {
    let sig_path = pkgdir.join("PKGBUILD.sig");
    let pkgb_path = pkgdir.join("PKGBUILD");

    // Use structured verification
    let result = verify_signature(&sig_path, &pkgb_path);

    match result {
        VerificationResult::Valid { key_id, .. } => {
            println!("[reap] gpg :: PKGBUILD signature verified (trusted)");
            if let Some(trust) = get_trust_level(&key_id) {
                println!("[reap] gpg :: Key {} trust level: {}", key_id, trust);
            }
            Ok(())
        }
        VerificationResult::ValidUntrusted { key_id, .. } => {
            println!(
                "[reap] gpg :: PKGBUILD signature valid but key not fully trusted"
            );
            if let Some(trust) = get_trust_level(&key_id) {
                println!("[reap] gpg :: Key {} trust level: {}", key_id, trust);
            }
            Ok(())
        }
        VerificationResult::MissingKey { key_id } => {
            println!("[reap] gpg :: Missing public key: {}", key_id);

            // Attempt auto key fetch
            let keyserver = "hkps://keys.openpgp.org";
            println!(
                "[reap] gpg :: Attempting to fetch key {} from {}...",
                key_id, keyserver
            );

            let fetch = Command::new("gpg")
                .args(["--keyserver", keyserver, "--recv-keys", &key_id])
                .status();

            match fetch {
                Ok(s) if s.success() => {
                    println!(
                        "[reap] gpg :: Successfully imported key {} from {}",
                        key_id, keyserver
                    );
                    // Re-run verification using verify_signature
                    let retry = verify_signature(&sig_path, &pkgb_path);
                    match retry {
                        VerificationResult::Valid { .. }
                        | VerificationResult::ValidUntrusted { .. } => {
                            println!(
                                "[reap] gpg :: PKGBUILD signature verified after key import"
                            );
                            Ok(())
                        }
                        _ => Err(format!(
                            "[reap] gpg :: Verification still failed after importing key {}",
                            key_id
                        )),
                    }
                }
                Ok(_) => Err(format!(
                    "[reap] gpg :: Failed to import key {} from {}",
                    key_id, keyserver
                )),
                Err(e) => Err(format!(
                    "[reap] gpg :: Error running gpg --recv-keys: {}",
                    e
                )),
            }
        }
        VerificationResult::Invalid { reason } => {
            Err(format!("[reap] gpg :: Invalid signature: {}", reason))
        }
        VerificationResult::NoSignature => {
            Err("[reap] gpg :: PKGBUILD or signature missing".to_string())
        }
        VerificationResult::Error(e) => {
            Err(format!("[reap] gpg :: Error: {}", e))
        }
    }
}

/// Refresh all GPG keys
pub fn refresh_keys() {
    let status = Command::new("gpg").arg("--refresh-keys").status();
    if let Ok(s) = status {
        if s.success() {
            println!("[reap] gpg :: Refreshed all keys");
        } else {
            eprintln!("[reap] gpg :: Failed to refresh keys");
        }
    }
}

/// Async GPG key import from multiple keyservers
pub async fn import_gpg_key_async(keyid: &str) -> Result<(), String> {
    let keyservers = [
        "hkps://keyserver.ubuntu.com",
        "hkps://keys.openpgp.org",
        "hkps://pgp.mit.edu",
    ];
    let mut last_err = None;
    for server in &keyservers {
        match TokioCommand::new("gpg")
            .args(["--keyserver", server, "--recv-keys", keyid])
            .status()
            .await
        {
            Ok(s) if s.success() => {
                println!("[reap] gpg :: Imported key {} from {}", keyid, server);
                return Ok(());
            }
            Ok(_) => {
                last_err = Some(format!("Failed to import key from {}", server));
            }
            Err(e) => {
                last_err = Some(format!("TokioCommand error for {}: {}", server, e));
            }
        }
    }
    Err(last_err.unwrap_or_else(|| "All keyserver attempts failed".to_string()))
}

/// Async GPG key presence check
pub async fn check_key(keyid: &str) {
    let output = TokioCommand::new("gpg")
        .args(["--list-keys", keyid])
        .output()
        .await;
    if let Ok(out) = output {
        if out.status.success() {
            println!("[reap] gpg :: GPG key {} is present.", keyid);
        } else {
            println!("[reap] gpg :: GPG key {} is NOT present.", keyid);
        }
    } else {
        println!("[reap] gpg :: Failed to check GPG key {}.", keyid);
    }
}

/// Show GPG key details - user-facing informational output
///
/// Wrapper for `show_gpg_key_info()`. Use for debugging or user inspection,
/// not for programmatic trust decisions.
pub fn show_key(keyid: &str) {
    show_gpg_key_info(keyid);
}

/// Check if GPG key exists in keyring
pub fn key_exists(keyid: &str) -> bool {
    let output = Command::new("gpg").args(["--list-keys", keyid]).output();
    if let Ok(out) = output {
        out.status.success()
    } else {
        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_verification_result_is_valid() {
        // Test Valid is recognized as valid
        let valid = VerificationResult::Valid {
            key_id: "ABC123".to_string(),
            fingerprint: "0123456789ABCDEF".to_string(),
        };
        assert!(valid.is_valid());
        assert!(valid.is_trusted());

        // Test ValidUntrusted is valid but not trusted
        let valid_untrusted = VerificationResult::ValidUntrusted {
            key_id: "ABC123".to_string(),
            fingerprint: "0123456789ABCDEF".to_string(),
        };
        assert!(valid_untrusted.is_valid());
        assert!(!valid_untrusted.is_trusted());

        // Test Invalid is not valid
        let invalid = VerificationResult::Invalid {
            reason: "test".to_string(),
        };
        assert!(!invalid.is_valid());
        assert!(!invalid.is_trusted());

        // Test MissingKey is not valid
        let missing = VerificationResult::MissingKey {
            key_id: "ABC123".to_string(),
        };
        assert!(!missing.is_valid());
        assert!(!missing.is_trusted());

        // Test NoSignature is not valid
        let no_sig = VerificationResult::NoSignature;
        assert!(!no_sig.is_valid());
        assert!(!no_sig.is_trusted());

        // Test Error is not valid
        let error = VerificationResult::Error("test".to_string());
        assert!(!error.is_valid());
        assert!(!error.is_trusted());
    }

    #[test]
    fn test_verification_result_variants_distinguishable() {
        // Ensure we can pattern match all variants
        let results = vec![
            VerificationResult::Valid {
                key_id: "A".to_string(),
                fingerprint: "B".to_string(),
            },
            VerificationResult::ValidUntrusted {
                key_id: "A".to_string(),
                fingerprint: "B".to_string(),
            },
            VerificationResult::Invalid {
                reason: "bad".to_string(),
            },
            VerificationResult::MissingKey {
                key_id: "A".to_string(),
            },
            VerificationResult::NoSignature,
            VerificationResult::Error("err".to_string()),
        ];

        let mut valid_count = 0;
        let mut trusted_count = 0;

        for r in &results {
            if r.is_valid() {
                valid_count += 1;
            }
            if r.is_trusted() {
                trusted_count += 1;
            }
        }

        assert_eq!(valid_count, 2); // Valid and ValidUntrusted
        assert_eq!(trusted_count, 1); // Only Valid
    }
}
